bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/handler/GreeHandler.java

```diff
@@ -0,0 +1,439 @@
+        }
+
+        if (mode != -1) {
+            // Turn on the unit if currently off
+            if (!isNumber && (device.getIntStatusVal("Pow") == 0)) {
+                logger.debug("Send Auto-ON for mode {}", mode);
+                device.setDevicePower(socket, 1);
+            }
+
+            // Select mode
+            logger.debug("Select mode {}", mode);
+            device.SetDeviceMode(socket, mode);
+
+            // Check for secondary action
+            switch (modeStr) {
+                case MODE_ECO:
+                    // Turn on power saving for eco mode
+                    logger.debug("Turn on Power-Saving");
+                    device.setDevicePwrSaving(socket, 1);
+                    break;
+                case MODE_TURBO:
+                    device.setDeviceTurbo(socket, 1);
+                    break;
+            }
+        }
+    }
+
+    private boolean isMinimumRefreshTimeExceeded() {
+        long currentTime = System.currentTimeMillis();
+        long timeSinceLastRefresh = currentTime - lastRefreshTime;
+        if (timeSinceLastRefresh < MINIMUM_REFRESH_TIME_MS) {
+            return false;
+        }
+        lastRefreshTime = currentTime;
+        return true;
+    }
+
+    private void startAutomaticRefresh() {
+
+        Runnable refresher = new Runnable() {
+            @Override
+            public void run() {
+
+                try {
+                    logger.debug("Executing automatic update of values");
+                    // safeguard for multiple REFRESH commands
+                    if (isMinimumRefreshTimeExceeded()) {
+                        logger.debug("Fetching status values from device.");
+                        // Get the current status from the Airconditioner
+                        device.getDeviceStatus(clientSocket.get());
+                    } else {
+                        logger.trace(
+                                "Skipped fetching status values from device because minimum refresh time not reached");
+                    }
+
+                    // Update All Channels
+                    List<Channel> channels = getThing().getChannels();
+                    for (Channel channel : channels) {
+                        publishChannel(channel.getUID());
+                    }
+                } catch (GreeException e) {
+                    if (!e.isTimeout()) {
+                        logger.debug("Unable to perform auto-update: {}", e.toString());
+                    }
+                } catch (RuntimeException e) {
+                    logger.debug("Unable to perform auto-update", e);
+                }
+            }
+        };
+
+        refreshTask = scheduler.scheduleWithFixedDelay(refresher, 0, config.refresh, TimeUnit.SECONDS);
+        logger.debug("Automatic refresh started ({} second interval)", config.refresh);
+    }
+
+    private void publishChannel(ChannelUID channelUID) throws GreeException {
```

Comment:
What throws a `GreeException` here?

Reply:
updateNumber() or updateOnOff()

Reply:
`GreeException` is never thrown, because it is catched in this method. So, the throws declaration can be removed.

Reply:
I agree with @fwolter here, the GreeException should be removed from the throws clause since you are already caching it in the body.

Reply:
removed